### PR TITLE
Basics for challenge_receptionist: step 1/N

### PR DESCRIPTION
**Pull request by LoyVanBeek**
I'd like to merge this as a first step. 

The core states aren't implemented yet but depend on https://github.com/tue-robotics/image_recognition/issues/87
So, before continuing, let's get that to work and then continue with person stuff.

### CODE REVIEW

File: challenge_receptionist/package.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
+<package>
```
Comment (MatthijsBurgh) [inline challenge_receptionist/package.xml]: Time to switch to format 2?

File: challenge_receptionist/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+catkin_python_setup()
+
+catkin_package()
+
+# ------------------------------------------------------------------------------------------------
+#                                              BUILD
+# ------------------------------------------------------------------------------------------------
+
+include_directories(
+    include
```
Comment (MatthijsBurgh) [inline challenge_receptionist/CMakeLists.txt]: There is no include dir in this package

File: robot_skills/src/robot_skills/mockbot.py [hunk 3]
```diff
@@ -236,25 +263,31 @@ class Mockbot(robot.Robot):
     >>> Mockbot()
     """
     def __init__(self, *args, **kwargs):
-        self.robot_name = "mockbot"
+        robot_name = "mockbot"
+
+        rospy.set_param("/" + robot_name + "/skills/arm/offset/grasp_offset", {'x': 0, 'y': 0, 'z': 0})
```
Comment (MatthijsBurgh) [inline robot_skills/src/robot_skills/mockbot.py]: These are now arm specific

File: robot_skills/src/robot_skills/mockbot.py [hunk 4]
```diff
@@ -14,21 +14,38 @@
 
 # TU/e Robotics
 import arms
-from dragonfly_speech_recognition.msg import Choice
-from dragonfly_speech_recognition.srv import GetSpeechResponse
+from robot_skills.robot_part import RobotPart
+# from dragonfly_speech_recognition.msg import Choice
```
Comment (MatthijsBurgh) [inline robot_skills/src/robot_skills/mockbot.py]: Remove commented code

File: challenge_receptionist/scripts/challenge_receptionist [hunk 5]
```diff
@@ -1,3 +1,10 @@
-#! /usr/bin/env python
+#!/usr/bin/python
```
Comment (MatthijsBurgh) [inline challenge_receptionist/scripts/challenge_receptionist]: Why? I would use `/usr/bin/env python`

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 6]
```diff
@@ -461,5 +459,6 @@ def execute(self, userdata=None):
         return 'succeeded'
 
 if __name__ == "__main__":
+    rospy.init_node('human_interaction_doctest')
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Why is this needed?

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 7]
```diff
@@ -414,6 +414,51 @@ def learn_person_procedure(robot, person_name="", n_samples=5, timeout=5.0):
...
+            return 'failed'
+        else:
+            try:
+                print answer.resolve()
+                name = answer.resolve().semantics["name"]
+                self.personNameDes.write(name)
+
+                rospy.loginfo("Result received from speech recognition is '" + name + "'")
+            except KeyError, ke:
+                rospy.loginfo("KeyError resolving the name heard: " + str(ke))
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Why catch this silently? Your challenge will crash later on and you'll have to find this line that was printed (as info) a while ago. If an exception occurs: have it fail.

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 8]
```diff
@@ -414,6 +414,51 @@ def learn_person_procedure(robot, person_name="", n_samples=5, timeout=5.0):
...
+        answer = ds.VariableDesignator(resolve_type=QueryResult)
+        state = HearOptionsExtra(self.robot, spec, answer.writeable)
+        outcome = state.execute()
+
+        if not outcome == "heard":
+            self.personNameDes.write(self.defaultName)
+
+            rospy.logwarn("Speech recognition outcome was not successful (outcome: '{0}'). Using default name '{1}'".format(str(outcome), self.personNameDes.resolve()))
+            return 'failed'
+        else:
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: I don't think you need this `else`

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+#! /usr/bin/env python
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: Shebang

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+# ROS
+import rospy
+import smach
+
+# TU/e Robotics
+from robot_skills.arms import Arm
+from robot_skills.util.entity import Entity
+from robot_smach_states.util.designators import check_type
+
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: Pep8

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                return "failed"
+        else:
+            look_at_ent = point_at_ent  # type: Entity
+
+        arm = self._arm_designator.resolve()  # type: Arm
+        if not arm:
+            rospy.logerr("Could not resolve _arm_designator")
+            return "failed"
+
+        # TODO: make arm point at some pose
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: TODO? What does `point_to` do?

File: challenge_receptionist/src/challenge_receptionist/receptionist.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+import robot_smach_states.util.designators as ds
+
+from robocup_knowledge import load_knowledge
+from robot_skills.util.entity import Entity
+from hmi_msgs.msg import QueryResult
+
+challenge_knowledge = load_knowledge('challenge_receptionist')
+
+
+class ChallengeReceptionist(smach.StateMachine):
```
Comment (jlunenburg) [inline challenge_receptionist/src/challenge_receptionist/receptionist.py]: Breaking this up in multiple sub-state-machines allows for easier re-use and easier testing

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py (hunk 9 above)
Comment (LoyVanBeek) [reply robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: What should it be instead?

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 13]
```diff
@@ -414,6 +414,51 @@ def learn_person_procedure(robot, person_name="", n_samples=5, timeout=5.0):
...
+        rospy.loginfo("AskPersonName")
+
+        self.robot.speech.speak("What is your name?", block=True)
+
+        names_spec = "T['name': N] -> NAME[N]\n\n"
+        for dn in self.name_options:
+            names_spec += "NAME['{name}'] -> {name}\n".format(name=dn)
+        spec = ds.Designator(names_spec)
+
+        answer = ds.VariableDesignator(resolve_type=QueryResult)
```
[the remark this answers is not in the document]
Comment (LoyVanBeek) [reply robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: I know, but copy-pasted this from elsewhere and it just works :man_shrugging: . Will fix later on.

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py (hunk 11 above)
Comment (LoyVanBeek) [reply robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: Point at some Entity.

File: robocup_knowledge/src/robocup_knowledge/environments/robotics_testlabs/challenge_receptionist.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
+from robocup_knowledge import knowledge_loader
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robocup_knowledge/src/robocup_knowledge/environments/robotics_testlabs/challenge_receptionist.py]: As discussed, this fix was not what I meant

File: robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py [hunk 15]
```diff
@@ -38,13 +40,15 @@ class Say(smach.State):
     >>> #robot.speech.speak.assert_any_call('a', 'us', 'kyle', 'default', 'excited', True)
     >>> #robot.speech.speak.assert_any_call('b', 'us', 'kyle', 'default', 'excited', True)
     >>> #robot.speech.speak.assert_any_call('c', 'us', 'kyle', 'default', 'excited', True)"""
-    def __init__(self, robot, sentence=None, language=None, personality=None, voice=None, mood=None, block=True, look_at_standing_person=False):
+
+    def __init__(self, robot, sentence=None, language=None, personality=None, voice=None, mood=None, block=True,
+                 look_at_standing_person=False):
         smach.State.__init__(self, outcomes=["spoken"])
         ds.check_type(sentence, str, list)
```
Comment (jlunenburg) [inline robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py]: Remove all commented code from this file?

File: robot_smach_states/src/robot_smach_states/manipulation/point_at.py (hunk 11 above)
Comment (jlunenburg) [reply robot_smach_states/src/robot_smach_states/manipulation/point_at.py]: So... Why is this ToDo still in place (is the `point_at` not what you mean)? Furthermore, can we removed the `send_joint_trajectory`?

**Comment by MatthijsBurgh** [comment]
Can this PR be closed as #835 is merged?
